pkg/distributor/distributor_ingest_storage_test.go

```diff
@@ -265,6 +268,65 @@ func TestDistributor_Push_ShouldSupportIngestStorage(t *testing.T) {
 	}
 }
 
+func TestDistributor_Push_ShouldReturnErrorMappedTo4xxStatusCodeIfWriteRequestContainsTimeseriesBiggerThanLimit(t *testing.T) {
+	ctx := user.InjectOrgID(context.Background(), "user")
+	now := time.Now()
+
+	hugeLabelValueLength := 100 * 1024 * 1024
+
+	createWriteRequest := func() *mimirpb.WriteRequest {
+		return &mimirpb.WriteRequest{
+			Timeseries: []mimirpb.PreallocTimeseries{
+				makeTimeseries([]string{model.MetricNameLabel, strings.Repeat("x", hugeLabelValueLength)}, makeSamples(now.UnixMilli(), 1), nil),
+			},
+		}
+	}
+
+	limits := prepareDefaultLimits()
+	limits.MaxLabelValueLength = hugeLabelValueLength
+
+	testConfig := prepConfig{
+		numDistributors:         1,
+		ingestStorageEnabled:    true,
+		ingestStoragePartitions: 1,
+		limits:                  limits,
+	}
+
+	distributors, _, regs, _ := prepare(t, testConfig)
+	require.Len(t, distributors, 1)
+	require.Len(t, regs, 1)
+
+	t.Run("Push()", func(t *testing.T) {
+		// Send write request.
+		res, err := distributors[0].Push(ctx, createWriteRequest())
+		require.Error(t, err)
+		require.Nil(t, res)
+
+		// We expect a gRPC error.
+		errStatus, ok := grpcutil.ErrorToStatus(err)
+		require.True(t, ok)
+		assert.Equal(t, codes.FailedPrecondition, errStatus.Code())
+		assert.ErrorContains(t, errStatus.Err(), ingest.ErrWriteRequestDataItemTooLarge.Error())
+
+		// We expect the gRPC error to be detected as client error.
+		assert.True(t, mimirpb.IsClientError(err))
+	})
+
+	t.Run("Handler()", func(t *testing.T) {
```

Comment:
Note to reviewers: added this test case to explicitly test the returned HTTP status code.